query.go

```
@@ -0,0 +1,166 @@
+// into each other. It is from a slice of strings, reflecting the structure of url.Values
+// These can be specified by their type (whichever struct the Parameter mapper will be,
+// and the restrictions defined on the type, defined by Validators slice below)
+type QueryParameterMapper interface {
+	Encode(interface{}) ([]string, error)
+	Decode([]string) (interface{}, error)
+}
+
+// Examples of mappers
+type StringQueryParameterMapper struct {
+	Validators []func(string) bool
+}
+
+func (sqpm StringQueryParameterMapper) Decode(src []string) (interface{}, error) {
+	if len(src) != 1 {
+		return nil, errors.New("expected only one value")
+	}
+
+	str := 	src[0]
+	for _, v := range sqpm.Validators {
+		if !v(str) {
+			return nil, errors.New("a validation test failed")
+		}
+	}
+	return str, nil
+}
+
+func (sqpm StringQueryParameterMapper) Encode(src interface{}) ([]string, error) {
+	s, ok := src.(string)
+	if !ok {
+		return nil, errors.New("improper type")
+	}
+
+	return []string{s}, nil
+}
+
+type IntQueryParameterMapper struct {
+	Validators []func(int) bool
+}
+
+func (iqpm IntQueryParameterMapper) Decode(src []string) (interface{}, error) {
+	if len(src) != 1 {
+		return nil, errors.New("expected only one value")
+	}
+
+	num, err := strconv.Atoi(src[0])
+	if err != nil {
+		return nil, errors.New("param could not be converted to integer")
+	}
+
+	for _, v := range iqpm.Validators {
+		if !v(num) {
+			return nil, errors.New("a validation test failed")
+		}
+	}
+	return num, nil
+}
+
+func (iqpm IntQueryParameterMapper) Encode(src interface{}) ([]string, error) {
+	n, ok := src.(int)
+	if !ok {
+		return nil, errors.New("improper type")
+	}
+	return []string{strconv.Itoa(n)}, nil
+}
+
+type StrSliceQueryParameterMapper struct {
+	Validators []func([]string) bool
+	UnderlyingQueryParameterMapper QueryParameterMapper
+}
+
+func (sqpm StrSliceQueryParameterMapper) Decode(src []string) (interface{}, error) {
+	for _, val := range sqpm.Validators {
+		if !val(src) {
```

should we return an error here?

---

Fixed